servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java

```diff
@@ -2224,11 +2224,37 @@ public final Publisher<T> takeWhile(Predicate<? super T> predicate) {
      *
      * @param until {@link Completable}, termination of which, terminates the returned {@link Publisher}.
      * @return A {@link Publisher} that only emits the items till {@code until} {@link Completable} is completed.
-     *
+     * @deprecated Use {@link #takeUntil(Supplier)}.
      * @see <a href="https://reactivex.io/documentation/operators/takeuntil.html">ReactiveX takeUntil operator.</a>
      */
+    @Deprecated
     public final Publisher<T> takeUntil(Completable until) {
-        return new TakeUntilPublisher<>(this, until);
+        return takeUntil(() -> until);
+    }
+
+    /**
+     * Takes elements until {@link Completable} is terminated successfully or with failure.
+     * <p>
+     * This method provides a means to take a limited number of results from this {@link Publisher} and in sequential
+     * programming is similar to:
+     * <pre>{@code
+     *     List<T> results = ...;
+     *     for (T t : resultOfThisPublisher()) {
+     *         if (isCompleted(until)) {
+     *             break;
+     *         }
+     *         takeResults.add(t);
+     *     }
+     *     return results;
+     * }</pre>
+     *
+     * @param untilSupplier {@link Supplier} that is invoked on each subscribe that provides a {@link Completable},
+     * termination of which, terminates the returned {@link Publisher}.
+     * @return A {@link Publisher} that only emits the items till {@code until} {@link Completable} is completed.
+     * @see <a href="https://reactivex.io/documentation/operators/takeuntil.html">ReactiveX takeUntil operator.</a>
+     */
+    public final Publisher<T> takeUntil(Supplier<? extends Completable> untilSupplier) {
```

Comment:
@idelpivnitskiy - I know we have had discussion about providing `Supplier` vs using `defer(() -> ..)` in the past. We could continue to assume folks know they should use `defer(() -> ..)` but in this case it was missed in internal usage and when `Completable` terminates the Publisher is also terminated, which may trigger re-subscribe...

Reply:
discussed offline ... `Publisher.defer(() -> ..)` can be used as a building block in this case. we can add followup docs/examples to clarify making operators that may retain state re-subscribe safe.